block/src/Pos/Block/Logic/VAR.hs

```diff
@@ -71,6 +72,9 @@ verifyBlocksPrefix blocks = runExceptT $ do
     adoptedBV <- GS.getAdoptedBV
     let dataMustBeKnown = mustDataBeKnown adoptedBV
 
+    -- Ensure every block is internally consistent.
+    _ <- withExceptT VerifyBlocksError $ forM_ blocks verifyBlock
```

Comment:
@volhovm @georgeee verification is here. Is this where it needs to be?

Also, right below this it looks like we do a bunch of traversals over the list of blocks. We could probably speed up verification (and therefore improve daedalus UX) but rolling them all into one traversal.

Reply:
Ah, I just noticed that you already have a little monoidal verification thing: `Serokell.Util.VerificationRes`, and that's it's already used in `Pos.Block.Pure`. I'll include this verification in there, which just seems to make good sense.

Reply:
I think it should be done right after we get message from network and propagate it. Also don't forget that you should call `verifyHeader` when you do `getHeaders >>= getBlocks`.

Reply:
Why? You want to load the expensive computation as early as possible? Why don't we do it when all of the other validation is done?

Reply:
Internal datatype consistency is the least expensive computation that can be possibly done, pure/component checks are 10x+ expensive.

Because if adversary A sends us broken headers H, we can later request H from the honest node B and it'll consider us a malicious agent. It's not like the "ban for trash requests" policy is implemented, but there are thoughts about it. Also networking layer won't communicate with node sending garbage for a while. So we'd better check our requests for internal consistency.